samtranslator/model/sam_resources.py

```diff
@@ -1848,28 +1848,62 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
+        elif profile_type == "AWS_SNS_TOPIC_POLICY":
+            generated_resources.append(
+                self._construct_sns_topic_policy(source, destination, profile_properties, dest_index, multi_dest)
             )
-        if profile_type == "AWS_SQS_QUEUE_POLICY":
-            generated_resources.append(self._construct_sqs_queue_policy(source, destination, profile_properties))
-        if profile_type == "AWS_SNS_TOPIC_POLICY":
-            generated_resources.append(self._construct_sns_topic_policy(source, destination, profile_properties))
-        if profile_type == "AWS_LAMBDA_PERMISSION":
+        elif profile_type == "AWS_LAMBDA_PERMISSION":
             generated_resources.extend(
-                self._construct_lambda_permission_policy(source, destination, profile_properties)
+                self._construct_lambda_permission_policy(
+                    source, destination, profile_properties, dest_index, multi_dest
+                )
             )
+        else:
+            raise InvalidResourceException(self.logical_id, f"Profile type {profile_type} is not supported")
+        return generated_resources
+
+    @cw_timer
+    def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
+        resource_resolver: ResourceResolver = kwargs["resource_resolver"]
+        original_template = kwargs["original_template"]
 
-        generated_logical_ids = [resource.logical_id for resource in generated_resources]
+        multi_dest = True
+        if isinstance(self.Destination, dict):
+            multi_dest = False
+            self.Destination = [self.Destination]
+
+        list_generated_resources: List[Resource] = []
+
+        for dest_index, dest in enumerate(self.Destination):
+            try:
+                destination = get_resource_reference(dest, resource_resolver, self.Source)
+                source = get_resource_reference(self.Source, resource_resolver, dest)
+            except ConnectorResourceError as e:
+                raise InvalidResourceException(self.logical_id, str(e))
+
+            generated_resource = self.generate_resource(source, destination, dest_index, multi_dest, resource_resolver)
+
+            self._add_connector_metadata(generated_resource, original_template, source, destination)
+            list_generated_resources.extend(generated_resource)
+
+        generated_logical_ids = [resource.logical_id for resource in list_generated_resources]
         replace_depends_on_logical_id(self.logical_id, generated_logical_ids, resource_resolver)
 
-        self._add_connector_metadata(generated_resources, original_template, source, destination)
-        if generated_resources:
-            return generated_resources
+        if list_generated_resources:
+            return list_generated_resources
 
-        # Should support all profile types
-        raise TypeError(f"Unknown profile policy type '{profile_type}'")
+        raise TypeError("The destination is empty")
```

Comment:
nit: This error message doesn't seem quite right. There are other scenarios where `list_generated_resources` can be empty list.

Reply:
Do you know what scenarios `list_generated_resources` can be empty list? Trying to figure out the right message here.

Reply:
I don't think this will ever happen if our profile is right. So if it happens, it should be an availability dip. We can raise something like unexpected error

Reply:
updates: an empty destination list will trigger this error. So I updated to `InvalidResourceException` because it's a template error.